utils/histogram.hh

```diff
@@ -318,6 +319,89 @@ public:
     }
 };
 
+/*!
+ * \brief A class for a histogram-based summary calculation.
+ *
+ * A summary is a histogram where each bucket holds some quantile.
+ * While a histogram typically holds values from the system start,
+ * a summary is defined over some duration (i.e., latencies in the last 10 seconds).
+ * To calculate a summary, we use two estimated-histograms, calculate their delta, and get the
+ * summary from that delta histogram.
+ *
+ */
+class summary_calculator {
+    std::vector<double> _quantiles = { 0.5, 0.95, 0.99};
+    std::vector<double> _summary = { 0, 0, 0};
+    time_estimated_histogram _previous_histogram;
+    time_estimated_histogram _current_histogram;
+public:
+    /*!
+     * \brief update the summary and histograms
+     *
+     * The update method is called every time tick.
+     * When done, _previous_histogram would equal _current_histogram
+     * and the _summary would contain the current _summary calculation
+     *
+     * The calculation is done in two stages. first, we determine what is
+     * the cutoff for each quantile, for example, assume that there are new 1000
+     * entries and the quantiles are 0.5, 0.95 and 0.99
+     * The cutoffs will be 500, 950, and 990. We reuse the _summary array
+     * to hold these values.
+     *
+     * Second, while coping the _current_histogram to the _previous_histogram,
+     * we collect the diffs. Each time we cross a cutoff value, we update the
+     * _summary with the bucket limit (i.e., the latency value).
+     *
+     * To continue the previous example, if the first 3 diffs had the values:
+     * 10, 300, 200. When reaching the third one, the total diff will be 510,
+     * and we set the summary[0] as the third bucket limit.
+     *
+     */
+    void update() noexcept {
+        auto new_entries = _current_histogram.count() - _previous_histogram.count();
+        if (new_entries <= 0) {
+            return;
+        }
```

Comment:
If the number of entries is lower, surely the result changes.

Reply:
Unless I missunderstand you comment Histogram's count monotonically grow, 
so it's alwasy true that:
``` _current_histogram.count() >= _previous_histogram.count()```

Reply:
Then why do you check for <=?

Reply:
equal should work just as well, I'll change to make it clearer